junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java

```diff
@@ -375,26 +389,41 @@ private static String descriptionFor(Executable executable) {
 		private SortedMap<Path, IOException> deleteAllFilesAndDirectories(FileOperations fileOperations)
 				throws IOException {
 
-			if (this.dir == null || Files.notExists(this.dir)) {
+			Path rootDir = this.dir;
+			if (rootDir == null || Files.notExists(rootDir)) {
 				return Collections.emptySortedMap();
 			}
 
 			SortedMap<Path, IOException> failures = new TreeMap<>();
 			Set<Path> retriedPaths = new HashSet<>();
-			tryToResetPermissions(this.dir);
-			Files.walkFileTree(this.dir, new SimpleFileVisitor<Path>() {
+			Path rootRealPath = rootDir.toRealPath();
+
+			tryToResetPermissions(rootDir);
+			Files.walkFileTree(rootDir, new SimpleFileVisitor<Path>() {
 
 				@Override
-				public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
-					if (!dir.equals(CloseablePath.this.dir)) {
+				public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+					LOGGER.trace(() -> "preVisitDirectory: " + dir);
+					if (isLink(dir)) {
+						delete(dir);
+						return SKIP_SUBTREE;
+					}
```

Comment:
:speech_balloon: This avoids following junctions and, potentially, deleting files outside of the temp dir. We need to delete them here because `postVisitDirectory` is not called after returning `SKIP_SUBTREE`.